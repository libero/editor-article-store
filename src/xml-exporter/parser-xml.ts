import xmldom from "xmldom";
import nwmatcher from "nwmatcher";

export function parseXML(xml: string) {

  const xmlDoc = new xmldom.DOMParser().parseFromString(xml, 'text/xml');

  xmlDoc.constructor.prototype.querySelector = function (selectors: string) {
    return nwmatcher({document: xmlDoc}).first(selectors, this);
  };

  xmlDoc.constructor.prototype.querySelectorAll = function (selectors: string) {
    return nwmatcher({document: xmlDoc}).select(selectors, this);
  };

  xmlDoc.createElement('a').constructor.prototype.matches = function (selectors: string) {
    return nwmatcher({document: xmlDoc}).match(this, selectors);
  };

  return xmlDoc;
}
