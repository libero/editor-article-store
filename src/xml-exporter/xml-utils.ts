import xmldom from "xmldom";
import nwmatcher from "nwmatcher";

export function parseXML(xml: string) {

  const xmlDoc = new xmldom.DOMParser().parseFromString(xml, 'text/xml');

  xmlDoc.constructor.prototype.querySelector = function (selectors: string) {
    return nwmatcher({document: xmlDoc}).first(selectors, this);
  };

  xmlDoc.constructor.prototype.querySelectorAll = function (selectors: string) {
    return nwmatcher({document: xmlDoc}).select(selectors, this);
  };

  const Element = xmlDoc.createElement('a').constructor;

  Element.prototype.matches = function (selectors: string) {
    return nwmatcher({document: xmlDoc}).match(this, selectors);
  };

  Element.prototype.hasAttribute = function (attributeName: string) {
    return nwmatcher({document: xmlDoc}).hasAttribute(this, attributeName);
  };

  Element.prototype.getAttribute = function (attributeName: string) {
    return nwmatcher({document: xmlDoc}).getAttribute(this, attributeName);
  };

  Element.prototype.querySelector = function (selectors: string) {
    return nwmatcher({document: xmlDoc}).first(selectors, this);
  };

  return xmlDoc;
}

export function clearNode(el: Element) {
  Array.from(el.childNodes).forEach(child => child.parentNode!.removeChild(child))
}
