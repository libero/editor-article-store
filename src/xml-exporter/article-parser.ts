import {Manuscript} from "../model/manuscript";
import { createTitleState } from '../model/title';
import {createBodyState} from "../model/body";
import {createAbstractState, createImpactStatementState} from "../model/abstract";
import {createAcknowledgementsState} from "../model/acknowledgements";
import {Article} from "../types/article";

import { parseXML } from "./xml-utils";
import {createRelatedArticleState} from "../model/related-article";

export function getArticleManuscript(article: Article): Manuscript {
  const xmlDoc = parseXML(article.xml);

  const title = xmlDoc.querySelector('title-group article-title') as Element;
  // // const keywordGroups = doc.querySelectorAll('kwd-group');
  const abstract: Element | undefined = Array.from(xmlDoc.querySelectorAll('abstract'))
    .find((el: Element) => !el.hasAttribute('abstract-type'));

  const impactStatement: Element | undefined = Array.from(xmlDoc.querySelectorAll('abstract'))
    .find((el: Element) => el.getAttribute('abstract-type') === 'toc');

  // // const authors = doc.querySelectorAll('contrib[contrib-type="author"]');
  // // const affiliations = doc.querySelectorAll('contrib-group:first-of-type aff');
  // // const references = doc.querySelectorAll('ref-list ref element-citation');
  // // const authorNotes = doc.querySelector('author-notes');
  const relatedArticles = xmlDoc.querySelectorAll('related-article');
  const acknowledgements: Element | null = xmlDoc.querySelector('ack');
  const body = xmlDoc.querySelector('body') as Element;

  return {
    title: createTitleState(title),
    abstract: createAbstractState(abstract),
    impactStatement: createImpactStatementState(impactStatement),
    acknowledgements: createAcknowledgementsState(acknowledgements),
    body: createBodyState(body),

    // keywordGroups: createKeywordGroupsState(Array.from(keywordGroups)),
    // authors: authorsState,
    // affiliations: createAffiliationsState(Array.from(affiliations)),
    // references: createReferencesState(Array.from(references)),
    relatedArticles: createRelatedArticleState(Array.from(relatedArticles)),
    // articleInfo: new ArticleInformation(doc.documentElement, authorsState),
    // journalMeta: {
  //   //   publisherName: getTextContentFromPath(doc, 'journal-meta publisher publisher-name'),
  //   //   issn: getTextContentFromPath(doc, 'journal-meta issn')
  //   // }
  } as Manuscript;
}
