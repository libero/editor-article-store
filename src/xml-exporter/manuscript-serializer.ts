import { EditorState } from 'prosemirror-state';
import { DOMSerializer } from 'prosemirror-model';
import xmldom from 'xmldom';

import { Article } from '../types/article';
import { Manuscript } from '../model/manuscript';
import { parseXML } from '../xml-exporter/xml-utils';
import { serializeAbstractState, serializeImpactStatementState } from '../model/abstract';
import { serializeTitleState } from '../model/title';
import { serializeAcknowledgementState } from '../model/acknowledgements';
import { serializeBodyState } from '../model/body';
import { serializeRelatedArticles } from '../model/related-article';
import { serializeAuthors } from '../model/person';
import { serializeAffiliations } from '../model/affiliation';
import { serializeKeywordGroups } from '../model/keyword';
import { serializeReferenceState } from '../model/reference';
import { serializeArticleInformaion } from '../model/article-information';

export function createXmlDomSerializer(editorState: EditorState): DOMSerializer {
    return DOMSerializer.fromSchema(editorState.schema);
}

export function serializeManuscriptSection(editorState: EditorState, document: Document): DocumentFragment {
    const serializer = createXmlDomSerializer(editorState);
    return serializer.serializeFragment(editorState.doc.content, { document });
}

export function serializeManuscript(article: Article, manuscript: Manuscript): string {
    const xmlDoc = parseXML(article.xml);

    serializeTitleState(xmlDoc, manuscript);
    serializeAbstractState(xmlDoc, manuscript);
    serializeImpactStatementState(xmlDoc, manuscript);
    serializeAcknowledgementState(xmlDoc, manuscript);
    serializeBodyState(xmlDoc, manuscript);
    serializeRelatedArticles(xmlDoc, manuscript);
    serializeAuthors(xmlDoc, manuscript);
    serializeAffiliations(xmlDoc, manuscript);
    serializeKeywordGroups(xmlDoc, manuscript);
    serializeReferenceState(xmlDoc, manuscript);
    serializeArticleInformaion(xmlDoc, manuscript);
    return new xmldom.XMLSerializer().serializeToString(xmlDoc);
}
