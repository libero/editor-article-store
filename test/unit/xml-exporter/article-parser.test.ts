import {readFileSync} from "fs";
import {join, resolve} from "path";
import {getArticleManuscript} from "../../../src/xml-exporter/article-parser";
import {EditorState} from "prosemirror-state";
import {get} from "lodash";

jest.mock('uuid', () => ({
  v4: () => 'unique_id'
}));

describe('ArticleParser', function () {
  it('should parse article XML', () => {
    const xml = readFileSync(resolve(join(__dirname, '../..', '/test-files/manuscript.xml'))).toString('utf8');
    const article = {
      articleId: "1",
      datatype: "text/xml",
      fileName: "manuscript.xml",
      version: "1.0",
      xml
    };
    const manuscript = getArticleManuscript(article);

    expect(manuscript.articleInfo).toEqual({
      _id: 'unique_id',
      articleDOI: "10.7554/eLife.00104",
      articleType: "Insight",
      copyrightStatement: "Â© 2019, Atherden and Smith",
      dtd: "1.2",
      elocationId: "e00104",
      licenseText: expect.any(EditorState),
      licenseType: "CC-BY-4",
      publicationDate: "2019-11-30",
      publisherId: "00104",
      subjects: ['Cell Biology', 'Genetics and Genomics'],
      volume: "8"
    });

    expect(manuscript.title).toBeInstanceOf(EditorState);
    expect(manuscript.title.doc.textContent)
      .toBe(`Example eLife Insight2 article with a complete2 reference 'front' section`);

    expect(manuscript.abstract).toBeInstanceOf(EditorState);
    expect(manuscript.abstract.doc.textContent)
      .toBe(`New studies examine how the different sub-structures in the sed nisi neque2, ornare id enim1 et, dapibus rutrum urna.`);

    expect(manuscript.impactStatement).toBeInstanceOf(EditorState);
    expect(manuscript.impactStatement.doc.textContent)
      .toBe(`New studies examine how the different sub-structures in the sed nisi neque2, ornare id enim1 et, dapibus rutrum urna.`);

    expect(manuscript.acknowledgements).toBeInstanceOf(EditorState);
    expect(manuscript.acknowledgements.doc.textContent)
      .toBe(`Acknowledgements section containing only paragraph type elements (no block content such as tables, figures and so on).`);

    expect(manuscript.body).toBeInstanceOf(EditorState);
    expect(manuscript.body.doc.textContent).toMatchSnapshot();

    expect(manuscript.keywordGroups).toEqual(expect.objectContaining({
      'author-keywords': {
        title: undefined,
        keywords: [
          {_id: "unique_id", content: expect.any(EditorState)},
          {_id: "unique_id", content: expect.any(EditorState)},
          {_id: "unique_id", content: expect.any(EditorState)},
          {_id: "unique_id", content: expect.any(EditorState)},
          {_id: "unique_id", content: expect.any(EditorState)}
        ],
        newKeyword: {_id: "unique_id", content: expect.any(EditorState)}
      },
      'research-organism': {
        title: 'Research organism',
        keywords: [{_id: "unique_id", content: expect.any(EditorState)}],
        newKeyword: {_id: "unique_id", content: expect.any(EditorState)}
      }
    }));

    expect(get(manuscript, 'keywordGroups.author-keywords.keywords.0.content.doc.textContent')).toBe('cerebellum');
    expect(get(manuscript, 'keywordGroups.author-keywords.keywords.1.content.doc.textContent')).toBe('climbing fiber');
    expect(get(manuscript, 'keywordGroups.author-keywords.keywords.2.content.doc.textContent')).toBe('reinforcement learning');
    expect(get(manuscript, 'keywordGroups.author-keywords.keywords.3.content.doc.textContent')).toBe('vitae');
    expect(get(manuscript, 'keywordGroups.author-keywords.keywords.4.content.doc.textContent')).toBe('Purkinje cells');
    expect(get(manuscript, 'keywordGroups.research-organism.keywords.0.content.doc.textContent')).toBe('Mouse');

    expect(manuscript.authors.length).toBe(5);
    expect(manuscript.authors[0]).toStrictEqual(expect.objectContaining({
      _id: 'unique_id',
      firstName: 'Fred',
      lastName: 'Atherden',
      suffix: '',
      isAuthenticated: true,
      orcid: '0000-0002-6048-1470',
      bio: expect.any(EditorState),
      email: 'f.atherden@elifesciences.org',
      isCorrespondingAuthor: true,
      affiliations: ['aff1', 'aff2'],
      hasCompetingInterest: true,
      competingInterestStatement: 'Is an employee of eLife. No other competing interests exist'
    }));

    expect(manuscript.authors[1]).toStrictEqual(expect.objectContaining({
      _id: 'unique_id',
      firstName: 'Jeanine',
      lastName: 'Smith',
      suffix: 'III',
      isAuthenticated: false,
      orcid: '',
      bio: expect.any(EditorState),
      email: '',
      isCorrespondingAuthor: false,
      affiliations: ['aff2'],
      hasCompetingInterest: true,
      competingInterestStatement: 'No competing interests declared'
    }));

    expect(manuscript.authors[2]).toStrictEqual(expect.objectContaining({
      _id: 'unique_id',
      firstName: 'Jack',
      lastName: 'London',
      suffix: 'III',
      isAuthenticated: false,
      orcid: '',
      bio: expect.any(EditorState),
      email: '',
      isCorrespondingAuthor: false,
      affiliations: ['aff2'],
      hasCompetingInterest: false,
      competingInterestStatement: ''
    }));

    expect(manuscript.authors[3]).toStrictEqual(expect.objectContaining({
      _id: 'unique_id',
      firstName: 'Mark',
      lastName: 'Twain',
      suffix: 'III',
      isAuthenticated: false,
      orcid: '',
      bio: expect.any(EditorState),
      email: '',
      isCorrespondingAuthor: false,
      affiliations: ['aff2'],
      hasCompetingInterest: false,
      competingInterestStatement: ''
    }));

    expect(manuscript.authors[4]).toStrictEqual(expect.objectContaining({
      _id: 'unique_id',
      firstName: 'Alexandr',
      lastName: 'Solzhenitsin',
      suffix: 'III',
      isAuthenticated: false,
      orcid: '',
      bio: expect.any(EditorState),
      email: '',
      isCorrespondingAuthor: false,
      affiliations: ['aff2'],
      hasCompetingInterest: true,
      competingInterestStatement: 'No competing interests declared'
    }));

    expect(get(manuscript, 'authors.0.bio.doc.textContent'))
      .toBe('Fred Atherden is in the Production Department, eLife Sciences, Cambridge, United Kingdoms Creative Commons Attribution License');
    expect(get(manuscript, 'authors.1.bio.doc.textContent'))
      .toBe('Jeanine Smith III is in the Department, University, City, Country');
    expect(get(manuscript, 'authors.2.bio.doc.textContent'))
      .toBe('Jeanine Smith III is in the Department, University, City, Country');
    expect(get(manuscript, 'authors.3.bio.doc.textContent'))
      .toBe('Jeanine Smith III is in the Department, University, City, Country');
    expect(get(manuscript, 'authors.4.bio.doc.textContent')).toBe('');

    expect(manuscript.affiliations).toEqual([
      {
        "_id": "aff1",
        "address": {
          "city": "Cambridge",
        },
        "country": "United Kingdom",
        "institution": {
          "name": "Production Department, eLife Sciences",
        },
        "label": "1",
      },
      {
        "_id": "aff2",
        "address": {
          "city": "City",
        },
        "country": "Country",
        "institution": {
          "name": "Department, University",
        },
        "label": "2",
      },
    ]);

    expect(manuscript.relatedArticles).toEqual([{
      "_id": "ra1",
      "articleType": "commentary-article",
      "href": "10.7554/eLife.42697"
    }, {
      "_id": "ra2",
      "articleType": "commentary-article",
      "href": "10.7554/eLife.00067"
    }]);

    expect(manuscript.references).toMatchSnapshot();

  });
});
