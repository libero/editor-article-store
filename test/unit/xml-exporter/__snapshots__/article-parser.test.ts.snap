// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleParser should parse article XML 1`] = `"Related research article  Srinivasan S, Kaplan CN, Mahadevan L. 2019. A multiphase theory for spreadingmicrobial swarms and films. eLife 8:e42697. doi: 10.7554/eLife.42697Related research article Zoolander D, Hansel, Mugatu J. 2019. Second related article title (if present). eLife 8:e00067. doi: 10.7554/eLife.00067Lorem ipsum dolor sit amet, consectetur adipiscing elit (; ). Etiam imperdiet tortor nec enim feugiat euismod . Etiam malesuada tristique quam non fringilla (). Curabitur sem neque, pretium et lorem ac, porta lobortis nisl. Curabitur suscipit semper sodales. Orci varius natoque penatibus et magnis dis parturient montes, nascetur ridiculus mus. Vivamus cursus faucibus felis, consequat rutrum neque vehicula eget (http://www.broadinstitute.org/igv/). Donec in neque et ligula tristique semper id sit amet quam. Proin varius, arcu posuere faucibus rhoncus, leo neque porttitor augue, sit amet cursus neque neque quis turpis. Maecenas dignissim aliquam orci, maximus scelerisque nisl imperdiet in. (; ; ).First entry (bullet)First sub-entry (bullet)Second sub-entry (bullet)First sub-entry (bullet)Second sub-entry (bullet)First sub-entry (bullet)Second sub-entry (bullet)Second entry (bullet) integer suscipit purus a semper dictum . Duis finibus elementum ex sit amet pulvinar (; ). Cras  nec est in sem iaculis fermentum (). Aenean sit amet turpis pulvinar, ornare nisl ac, porta nisl. Vivamus sit amet maximus elit (Figure 1). Duis a ultrices augue. Duis id ipsum iaculis, imperdiet leo a, cursus purus ().Etiam nunc eros, porttitor tincidunt libero eu (), varius pretium nibh (; ; ). Curabitur nibh libero, gravida eu ex vel, rhoncus finibus est. Aliquam viverra augue vel posuere tempus (proin pharetra posuere eros vitae semper []). Duis aliquam ac ex non tristique. Nulla dignissim vel ipsum at vulputate (; ). Nam orci ipsum, dapibus ac sagittis non, blandit non enim ().Phasellus ac nisl a erat congue fringilla/fringilla. Maecenas nulla leo , viverra ut tortor at, tristique maximus urna. Aenean commodo mauris sapien, a vulputate sapien porttitor nec (see Figures 1 and 2). Integer id aliquam libero. Duis euismod velit augue, id ultricies enim varius sit amet.Integer in dui quis velit iaculis rhoncus (European Galaxy Server https://usegalaxy.eu). Vestibulum nec nulla eget nisi  cursus dignissim non sed felis (donec tincidunt vulputate lectus non malesuada [; ]). Nulla condimentum porttitor arcu a posuere. Duis dictum diam sit amet mauris elementum luctus. Mauris placerat arcu ut lobortis ultricies. Nunc ultricies, ante eget mollis viverra, elit urna feugiat nibh, malesuada dictum nulla est at lorem. Proin id ligula ligula. Nulla eu risus porta, ornare eros id, sodales risus(; ). Proin id metus in neque rutrum venenatis eget sed ante (). Pellentesque nec justo a orci sodales dapibus. Etiam quis felis feugiat, lobortis dui semper, imperdiet est.The resolution of a microscope, , is defined as the shortest distance between two points on a specimen that can still be seen as two distinct points in the image. In 1873 Ernst Abbe showed that diffraction limited the resolution of a microscope to , where is the wavelength of the light, is the refractive index of the medium the light travels through, and is the half-angle at which the objective captures light from the sample: is also known as the \\"numerical aperture\\" of the objective.Cras at diam eu mauris scelerisque fermentum . Sed semper mauris eget nulla faucibus, sed ultricies ligula pellentesque. Donec posuere mi non quam dapibus egestas scelerisque non magna. Vestibulum sit amet bibendum tortor. Interdum et malesuada fames ac ante ipsum primis in faucibus (; ; ).Praesent lobortis tempus nisi ac pulvinar (). Praesent at nulla ac lorem faucibus finibus. Ut vel maximus tortor, in dictum arcu. Proin condimentum felis eu dignissim posuere. Vivamus a tincidunt justo, eu eleifend mauris. Mauris porttitor erat eu purus dignissim, sed consectetur odio pharetra. Integer a lorem commodo, tempus nulla sed, vehicula metus. Nunc laoreet erat sed nibh iaculis aliquam. Nullam maximus ligula eu sem dictum dignissim.First entry (ordered)First sub-entry (bullet)Second sub-entry (bullet)First sub-entry (bullet)Second sub-entry (bullet)First sub-entry (bullet)Second sub-entry (bullet)Second entry (ordered)DisclaimerThe content of this publication does not necessarily reflect the views or policies of the Department of Health and Human Services, nor does mention of trade names, commercial products, or organizations imply endorsement by the US Government.Second level headingThe content of the second level sectionThird level headingThe content of the third level sectionFourth level headingThe content of the fourth level section"`;
